### PR TITLE
Fix interim fields result and calculation updating on the 2-Dimensional-CSV instrument

### PR DESCRIPTION
**Pull request by Lunga001**
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/719

## Current behavior before PR
      When there's only interim fields on the import file only 1 interims fields results is updated
      and the calculation of the analysis that has the interim fields does not get updated
## Desired behavior after PR is merged
      Update all interim field and the calculation of the analysis that has the interim fields
--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/exportimport/instruments/generic/two_dimension.py [hunk 1]
```diff
@@ -193,13 +240,27 @@ def parse_resultsline(self, line):
...
                     del quantitation['resultValue']
+                    for interim_res in list_of_interim_results:
+                        if kw in interim_res:
+                            # Interim field already in quantitation dict
+                            found = True
+                            break
+                    if found:
+                        continue
+                    interims = find_analysis_interims(quantitation['AR'])
+                    letscallitxxfornow = dict(zip(self._keywords, clean_splitted))
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/generic/two_dimension.py]: funny, but please change to something more meaningful;)

File: buildout.cfg [hunk 2]
```diff
@@ -30,6 +30,7 @@ parts =
 
 eggs =
     senaite.core
+    senaite.lims
```
Comment (ramonski) [inline buildout.cfg]: please remove from this PR

**Comment by Lunga001** [comment]
Done requested changes, sorry about that :)

File: bika/lims/exportimport/instruments/generic/two_dimension.py [hunk 3]
```diff
@@ -93,29 +93,63 @@ def is_keyword(kw):
...
+    keyword = None
+    analyses = find_analyses(ar_or_sample)
+    for analysis in analyses:
+        interims = []
+        if hasattr(analysis, 'getInterimFields'):
+            interims = analysis.getInterimFields()
+        for interim in interims:
+            if interim['keyword'] == kw:
+                keyword = analysis.getKeyword()
+                break
```
Comment (xispa) [inline bika/lims/exportimport/instruments/generic/two_dimension.py]: Note that you force a break only from inside the inner loop, but the outer loop (analyses) will continue unnecessarily.
The complexity of this function can be reduced as follows:

```python
def find_kw(ar_or_sample, kw):
    for analysis in find_analyses(ar_or_sample):
        interims = api.safe_getattr(analysis, 'getInterimFields', default=[])
        keywords = map(lambda item['keyword'], interims)
        if kw in keywords:
            return analysis.getKeyword()
    return None
```
Given the fact that the two first lines from within the loop are duplicated in `find_analysis_interims` I'd also suggest to create a function `get_interims_keywords`, so the complexity could be reduced a bit more::

```python
def get_interims_keywords(analysis):
    keywords = api.safe_get_attr(analysis, 'getInterimFields', default=[])
    return map(lambda item['keyword'], interims)

def find_analysis_interims(ar_or_sample):
    for analysis in find_analyses(ar_or_sample):
        keywords = get_interims_keywords(analysis)
        interim_fields.extend(keywords)
    return list(set(interim_fields))

def find_kw(ar_or_sample, kw):
    for analysis in find_analyses(ar_or_sample):
        if kw in get_interims_keywords(analysis):
            return analysis.getKeyword()
    return None
```

File: bika/lims/exportimport/instruments/generic/two_dimension.py [hunk 4]
```diff
@@ -93,29 +93,63 @@ def is_keyword(kw):
...
+        This function and is is_keyword function should probably be in
+        resultsimport.py or somewhere central where it can be used by other
+        instrument interfaces.
+    """
+    interim_fields = []
+    analyses = find_analyses(ar_or_sample)
+    for analysis in analyses:
+        interims = []
+        if hasattr(analysis, 'getInterimFields'):
+            interims = analysis.getInterimFields()
```
[the remark this answers is not in the document]
Comment (Lunga001) [reply bika/lims/exportimport/instruments/generic/two_dimension.py]: Done

File: bika/lims/exportimport/instruments/generic/two_dimension.py [hunk 5]
```diff
@@ -93,29 +93,63 @@ def is_keyword(kw):
...
+        for interim in interims:
+            if interim['keyword'] not in interim_fields:
+                interim_fields.append(interim['keyword'])
+    return interim_fields
```
[the remark this answers is not in the document]
Comment (Lunga001) [reply bika/lims/exportimport/instruments/generic/two_dimension.py]: Done

File: bika/lims/exportimport/instruments/generic/two_dimension.py (hunk 3 above)
Comment (Lunga001) [reply bika/lims/exportimport/instruments/generic/two_dimension.py]: Done

**Pull request merged**